### DISCUSSION TITLE
hit sound delay when playing with bluetooth headphones on linux

### DISCUSSION THREAD

**Discussion by PigeonCoding** [ppy/osu#25801] [Q&A]
### Type

Game behaviour

### Bug description

when playing osu with a bluetooth audio device (in linux at least i don't have a windows pc with bluetooth to test) the music audio is fine but the hit sfx has a very big delay more than the music itself which i found strange(aka the music as little to no delay apart from bluetooth of course but the hit delay is maybe around 1 sec if i had to guess which is a lot )

### Screenshots or videos

i unfortunately don't have a pc good enough to record and i don't even know if it would show in a recording

### Version

2023.1130.0

### Logs

[performance.log](https://github.com/ppy/osu/files/13696491/performance.log)
[network.log](https://github.com/ppy/osu/files/13696492/network.log)
[input.log](https://github.com/ppy/osu/files/13696493/input.log)
[runtime.log](https://github.com/ppy/osu/files/13696494/runtime.log)
[database.log](https://github.com/ppy/osu/files/13696495/database.log)

**Comment by peppy** [comment]
bluetooth has latency

**Reply by PigeonCoding** [reply]
i know but there should be no difference between the hit sfx and the music that's what i find weird since bluetooth latency is system wide

hit sfx is heard way after the circle is hit and i'm talking about more than 1s of latency and the problem is not the headphones because the music is relatively in sync the issue is the hit sfx

**Reply by peppy** [reply]
think about it for a moment. i can't explain this to you but it's how latency works.

**Reply by PigeonCoding** [reply]
https://youtu.be/H8ur7OjoLV0

Notice how the music is in sync but the hit sfx is delayed more than it should

And i know the concept of lag and latency in wireless communication like bluetooth and wifi

**Reply by Zyfarok** [reply]
@PigeonCoding are you sure you did not change the offset, either in the global settings or on the map ?

**Reply by PigeonCoding** [reply]
It is stock osu i just downloaded it and signed in it's the osu recommanded map that automatically downloads when you first launch it

**Reply by yellowbadbeast** [reply]
I've noticed that the offset tends to be a bit off on linux with stock settings, so it's probably just that.

**Reply by PigeonCoding** [reply]
Maybe ill investigate when i have more time

But i noticed this issue 2-3 months ago when i played osu with my bluetooth earphones

**Reply by n0Oo0Oo0b** [reply]
> hit sfx is heard way after the circle is hit and i'm talking about more than 1s of latency and the problem is not the headphones because the music is relatively in sync the issue is the hit sfx

If the music is in sync with the circles, that means the music is being played back before the hitcircle to compensate for the latency. For hitisounds to also be in sync with the music, they have to be played before they should be heard, which is impossible because they're based on your inputs.

**Reply by FlySlime** [reply]
I'll just hop on here shortly to report that I also experience this with Bluetooth speakers.

I've tried both stable and McOsu and they don't suffer from this issue (I use Linux). There is no extremely noticeable delay between the hit sounds and music. Of course as ppy mentions there will always be some delay with bluetooth output devices, however this sticks out way too much. So I suspect that there must be something going on with Lazer.

I'll check if the issue persists across different setups: Windows (as OP didn't have a Windows install), Linux X11, and Linux Wayland. Afterwards I'll report back.

**Reply by FlySlime** [reply]
After trying my bluetooth speakers on Linux and Windows I notice the following: 
- The issue seems to only persist on Linux installs, despite using X11 or Wayland. It seemed a little bit worse on Wayland however. 
- It is also not a Desktop Environment issue as both GNOME and KDE has this issue.
- In Windows there was no issue at all.

**Comment by smoogipoo** [comment]
There's nothing really actionable here at this stage - I have my own theories because of other experiences I've had, but testing is not easy.

That said, if you're up for compiling your own osu!framework, it would be interesting to test https://github.com/ppy/osu-framework/pull/6002

**Reply by PigeonCoding** [reply]
Unfortunately i don't have a lot of free time right now so i might have to wait a bit before doing it but it would be interesting to compile it

**Reply by Stoppedpuma** [reply]
Will add that this PR solves my only reason I don't use lazer for more than casual play.
